### ISSUE TITLE
Parameterized twins: learn auditable trait updates from accumulated prediction runs

### ISSUE THREAD

**Issue by johnjosephhorton** [expectedparrot/zwill#1]
## Summary

Implement a first-class Zwill workflow for **parameterized digital twins** whose respondent representations can improve from observed prediction errors without inserting target answers into persona text.

The workflow should:

1. declare and register a task-specific, interpretable trait schema;
2. populate a versioned trait vector for each respondent from permissible evidence;
3. render those traits through a shared persona template;
4. export EDSL jobs that elicit choice probabilities;
5. import and score predictions against observed training choices;
6. fit a conventional differentiable surrogate to accumulated `traits → LLM probabilities` observations;
7. use its gradients to propose bounded, inspectable trait diffs;
8. export verification jobs against the actual LLM;
9. accept, reject, or shrink diffs based on realized improvement; and
10. add every verification run to the growing training table before refitting.

Zwill should orchestrate and preserve these artifacts. It should **not execute EDSL directly**: Zwill builds `.ep` jobs, the user runs them with Expected Parrot, and Zwill imports the returned results.

## Motivation

Current twins are primarily prompt/persona artifacts. When a twin predicts the wrong option, Zwill has no structured way to propose a representation update. An LLM could suggest edits, but hidden prompt mutation is difficult to audit and can simply encode the answer.

A parameterized representation provides explicit bounded coordinates, a common template across respondents, auditable diffs, conventional models for estimating local response gradients, verification against the real LLM response surface, and an accumulating dataset across runs.

The prototype soft-drink experiment used eight brand-neutral traits and four choices. A ridge multivariate-logit surrogate was fitted to 60 trajectory observations. A single bounded gradient update was verified with 10 new LLM predictions: realized Brier loss fell from 0.7184 to 0.5388, mean probability on the observed choice rose from 0.289 to 0.392, and all 10 proposed updates improved Brier loss. This is proof of mechanism on synthetic, in-sample targets—not validation of respondent-level twins.

### 1. Trait schema

A declared, versioned JSON document:

```json
{
  "schema_id": "soft_drink_v1",
  "domain": "soft-drink choice",
  "bounds": [-1.0, 1.0],
  "traits": [
    {
      "name": "citrus",
      "group": "taste_preferences",
      "description": "Preference for bright citrus flavor",
      "presentation": {
        "negative": "tends not to enjoy citrus flavors",
        "neutral": "has no strong citrus preference",
        "positive": "tends to enjoy citrus flavors"
      }
    },
    {
      "name": "avoids_caffeine",
      "group": "constraints",
      "description": "Strength of caffeine avoidance"
    }
  ],
  "context_rules": [
    "With spicy food, prioritizes crispness",
    "In the evening, avoids caffeinated options"
  ],
  "forbidden_target_features": [
    "preferred_brand",
    "prefers_pepsi",
    "observed_choice"
  ]
}
```

Schemas may be authored manually or proposed by an exported LLM schema-generation job. Either way, Zwill should require review/approval and freeze the schema before respondent fitting.

Trait components are intentionally **context-dependent**. Soda choice calls for flavor and caffeine dimensions; another prediction task should use a different reviewed coordinate system.

### 2. Respondent trait state

Store one bounded vector per respondent plus provenance:

```json
{
  "respondent_id": "r_001",
  "schema_id": "soft_drink_v1",
  "revision": 0,
  "values": {
    "citrus": 0.7,
    "caramel": -0.2,
    "spiced": 0.1,
    "sweetness": 0.3,
    "avoids_caffeine": 0.8,
    "price_sensitivity": 0.4,
    "brand_loyalty": 0.2,
    "novelty_seeking": 0.5
  },
  "evidence": {
    "avoids_caffeine": ["training_response_q12", "training_response_q19"]
  }
}
```

The held-out target must never appear in trait evidence or rendered persona context.

### 3. Persona template

A schema-level template converts numeric values into consistent natural-language descriptions. Preserve the rendered persona and input state in the job manifest. Templates must not hard-code option or brand preferences. Leakage checks should inspect schema fields, evidence references, template text, and rendered personas.

### 4. Accumulating observation table

Each imported prediction appends an immutable row containing:

```text
run_id, respondent_id, schema_id, revision, trait_vector,
question_id, option_order, llm_probabilities, model, prompt_hash,
observed_choice_if_training, split, timestamp
```

Later runs add data rather than replacing earlier observations.

### 5. Surrogate model

Start with a transparent baseline:

- transform K probabilities into K−1 logits against a fixed reference option;
- fit ridge-regularized multivariate linear regression from standardized traits to logits;
- reconstruct probabilities with softmax;
- persist coefficients, intercepts, scaling parameters, regularization, option order, training row IDs, diagnostics, and code/version metadata.

Allow later nonlinear surrogates without changing upstream artifact contracts. Diagnostics should include row count/design coverage, leave-one-trajectory-out probability MAE, coefficients, predicted-versus-realized verification improvement, and warnings for rank deficiency, weak perturbation coverage, and extrapolation.

### 6. Proposed trait diff

For training observation `(x, y)`, optimize cross-entropy or Brier loss through the surrogate:

```text
x_candidate = clip(
    x - learning_rate * gradient(loss(surrogate(x), y)),
    schema_lower_bound,
    schema_upper_bound
)
```

The diff is data, not a hidden prompt mutation:

```json
{
  "respondent_id": "r_001",
  "from_revision": 0,
  "proposed_revision": 1,
  "objective": "cross_entropy",
  "learning_rate": 0.25,
  "changes": {
    "citrus": {"from": 0.70, "to": 0.61},
    "spiced": {"from": 0.10, "to": 0.18}
  },
  "predicted_loss_before": 1.42,
  "predicted_loss_after": 1.19,
  "surrogate_model_id": "surrogate_round_0"
}
```

Support per-coordinate and total-step bounds, immutable revision history, and accept/reject status.

### 7. Verification and accumulation

Zwill exports jobs for baseline and candidate states; the external runner executes them; Zwill imports the actual LLM distributions. Report predicted and realized changes in Brier loss and NLL, change in probability assigned to truth, top-choice changes, and whether each proposal improved or worsened.

Accepted and rejected candidates both contain information about the response surface and should be eligible for the next refit. Preserve round membership so diagnostics can avoid leakage across rounds.

## Suggested CLI shape

Exact names are open for discussion, but the workflow should be explicit:

```bash
# Declare and approve the representation
zwill traits schema register --survey soda --path traits.schema.json
zwill traits schema inspect --survey soda --schema soft_drink_v1
zwill traits schema approve --survey soda --schema soft_drink_v1 --approved-by "Reviewer"

# Populate states using permissible evidence
zwill traits populate-plan --survey soda --schema soft_drink_v1 --path populate-plan.json
zwill edsl build --survey soda --target trait-population-job \
  --plan populate-plan.json --path trait-population-jobs.ep
ep run trait-population-jobs.ep --output trait-population-results.ep
zwill traits population import --survey soda --input-path trait-population-results.ep

# Predict
zwill traits prediction-plan --survey soda --schema soft_drink_v1 \
  --heldout-question soda_choice --path prediction-plan.json
zwill edsl build --survey soda --target parameterized-twin-probability-job \
  --plan prediction-plan.json --path predictions.ep
ep run predictions.ep --output prediction-results.ep
zwill traits predictions import --survey soda --input-path prediction-results.ep

# Fit and propose updates
zwill traits surrogate fit --survey soda --schema soft_drink_v1 \
  --model ridge-multinomial-logit --path surrogate-round-0
zwill traits diff propose --survey soda --surrogate surrogate-round-0/model.json \
  --objective nll --learning-rate 0.25 --path diffs-round-1.jsonl

# Verify with the authoritative LLM
zwill edsl build --survey soda --target trait-diff-verification-job \
  --diffs diffs-round-1.jsonl --path verification-round-1.ep
ep run verification-round-1.ep --output verification-round-1-results.ep
zwill traits diff verify --survey soda --diffs diffs-round-1.jsonl \
  --input-path verification-round-1-results.ep --path verification-round-1

# Refit with all accumulated observations
zwill traits surrogate fit --survey soda --schema soft_drink_v1 \
  --through-round 1 --path surrogate-round-1
```

Every Zwill command should return structured `next_steps`. Prediction counts and model choices must be displayed and approved before export, consistent with the existing guide.

## Guardrails and validity

This spans two modes that must not be conflated:

1. **Training/fitting mode** may use a respondent's observed training choice to update that respondent's traits.
2. **Validation mode** must evaluate the frozen representation and fitting procedure on different held-out questions, choices, respondents, or a predeclared split.

Required safeguards:

- prevent target answers and target-revealing evidence from entering personas;
- freeze schemas/templates before respondent fitting;
- record prompts, models, hashes, inputs, outputs, diffs, and approvals;
- separate surrogate-fit rows from final evaluation rows;
- compare the final twin against uniform, marginal, one-shot, and conditional baselines;
- retain existing leakage audits and validation gates;
- clearly label oracle/in-sample optimization;
- do not claim success from top-1 accuracy alone;
- warn when repeated updates overfit one observed choice;
- make seeds, perturbation design, bounds, and stopping criteria explicit.

### Phase 1: contracts and manual traits

- schema registration, validation, inspection, and approval;
- respondent state import/export and revision history;
- deterministic persona rendering;
- parameterized probability-job builder and result import;
- observation-table export.

### Phase 2: linear surrogate and diffs

- ridge multivariate-logit fit;
- coefficient and approximation diagnostics;
- bounded gradient diff generation;
- verification job builder and realized-improvement report;
- append verified observations and refit.

### Phase 3: generated schemas/population and validation

- exported LLM jobs for context-dependent schema proposals and trait population;
- evidence provenance and leakage review;
- cross-fitting and held-out evaluation;
- integration into validation bundles and generated reports;
- optional nonlinear surrogates and trust-region update policies.

## Acceptance criteria for the first end-to-end release

- [ ] Register and approve a bounded trait schema without target-option traits.
- [ ] Import or externally generate an `AgentList`-compatible respondent state.
- [ ] Deterministically render and audit templated personas.
- [ ] Export—but do not execute—EDSL probability jobs.
- [ ] Append reproducible observation rows from imported results.
- [ ] Fit and serialize the ridge multivariate-logit surrogate.
- [ ] Expose coefficients and local approximation diagnostics.
- [ ] Propose inspectable, bounded, revisioned trait diffs.
- [ ] Export verification jobs and import realized LLM outcomes.
- [ ] Compare surrogate-predicted and realized loss changes.
- [ ] Accumulate verified runs into later fits with lineage preserved.
- [ ] Test option ordering, clipping, bounds, revision lineage, parsing, and leakage failures.
- [ ] Document the soft-drink example and distinguish fitting from held-out evaluation.

## Prototype artifacts

The working prototype and note currently live under:

```text
experiments/soft_drink_twins/
  fit_surrogate_gradient.py
  evaluate_surrogate_gradient.py
  surrogate_observations.jsonl
  surrogate_model_round0.json
  surrogate_model_round1.json
  surrogate_gradient_diffs_round1.jsonl
  surrogate_gradient_evaluation_round1.json
  twin-fit-tutorial.html
  methodological-note.tex
  methodological-note.pdf
```

These should inform contracts and tests, but production implementation should follow Zwill's managed-state, approval, artifact, remote-execution, leakage, and validation conventions.
